File: 03链表逆序/Solution.cpp
// 输入一个链表，按链表从尾到头的顺序返回一个ArrayList。

class Solution {
public:
    vector<int> printListFromTailToHead(ListNode* head) {
        vector<int>res;
        stack<ListNode*>node;
        ListNode* pNode=head;
        while(pNode!=nullptr) {
            node.push(pNode);
            pNode=pNode->next;
        }
        while(!node.empty()) {
            pNode=node.top();
            node.pop();
            res.push_back(pNode->val);
        }
        return res;
    }
};
/* cpp的栈空间解法 注意cpp的空指针定义和栈定义和java的区别
有一点区别需要注意 cpp的pop是没有返回值的，所以需要使用 
top方法将栈顶元素提取出来
*/

File: 03链表逆序/Solution.java
import java.util.ArrayList;
public class Solution {
    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        ArrayList<Integer> res =new ArrayList<>();
        if (listNode!=null){
            res.addAll(printListFromTailToHead(listNode.next));
            res.add(listNode.next);
        }
        return res;
    }
}
/* 方法二，使用递归，一直走到链表的最后一个值，然后逐级返回，一边
返回一边将当前链表值存入arraylist */